### PR TITLE
[Duboku] Decrypt m3u8 url link

### PR DESCRIPTION
**Pull request by DmitryScaletta**
### Description of your *pull request* and other information

`player_data`

```json
{
  "flag": "play",
  "encrypt": 2,
  "trysee": 0,
  "points": 0,
  "link": "/vodplay/4283-1-1.html",
  "link_next": "",
  "link_pre": "/vodplay/4283-1-31.html",
  "url": "JTY4JTc0JTc0JTcwJTczJTNBJTJGJTJGJTczJTc0JTZEJTJFJTZFJTYyJTZGJTZCJTc1JTJFJTYzJTZGJTZEJTJGJTMyJTMwJTMyJTM0JTMwJTMyJTMwJTM3JTJGJTQ0JTRDJTcxJTM0JTQyJTUxJTQ1JTVBJTJGJTY5JTZFJTY0JTY1JTc4JTJFJTZEJTMzJTc1JTM4",
  "url_next": "",
  "from": "vidjs",
  "server": "no",
  "note": "",
  "id": "4283",
  "sid": 1,
  "nid": 32
}
```

https://u.duboku.io/static/js/player.js

```js
if (player_data.encrypt == "1") {
  player_data.url = unescape(player_data.url);
  player_data.url_next = unescape(player_data.url_next);
} else if (player_data.encrypt == "2") {
  player_data.url = unescape(base64decode(player_data.url));
  player_data.url_next = unescape(base64decode(player_data.url_next));
}
```

Fixes https://github.com/yt-dlp/yt-dlp/issues/9159

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check all of the following options that apply:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Fix or improvement to an extractor (Make sure to add/update tests)
- [ ] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

</details>

### CODE REVIEW

File: yt_dlp/extractor/duboku.py [hunk 1]
```diff
@@ -129,11 +131,14 @@ def _real_extract(self, url):
         data_url = player_data.get('url')
         if not data_url:
             raise ExtractorError('Cannot find url in player_data')
-        data_from = player_data.get('from')
+        if player_data.get('encrypt') == 1:
+            data_url = urllib.parse.unquote(data_url)
+        if player_data.get('encrypt') == 2:
+            data_url = urllib.parse.unquote(base64.b64decode(data_url))
```
Comment (Grub4K) [inline yt_dlp/extractor/duboku.py:137]: Unquote only works on bytes in Python >=3.9 ([ref](<https://docs.python.org/3/library/urllib.parse.html#urllib.parse.unquote>))
```suggestion
            data_url = urllib.parse.unquote(base64.b64decode(data_url).decode('ascii'))
```

**Review by Grub4K — CHANGES_REQUESTED** [review]

**Review by Grub4K — APPROVED** [review]

**Comment by dirkf** [comment]
Download tests are failing with 401 from the UK on getting the "m3u8 information": is that expected?

**Comment by DmitryScaletta** [comment]
They use Cloudflare DDoS Protection now.
Extraction no longer works.
But it works fine with `User-Agent` header and `cf_clearance` cookie.

**Comment by coletdjnz** [comment]
https://github.com/yt-dlp/yt-dlp/pull/7595 may help with that

**Comment by bashonly** [comment]
I'm still getting 403'd with curl_cffi/curl-impersonate

**Review by seproDev — APPROVED** [review]
No longer has a CF challenge for me

File: yt_dlp/extractor/duboku.py [hunk 2]
```diff
@@ -129,11 +131,14 @@ def _real_extract(self, url):
         data_url = player_data.get('url')
         if not data_url:
             raise ExtractorError('Cannot find url in player_data')
-        data_from = player_data.get('from')
+        if player_data.get('encrypt') == 1:
+            data_url = urllib.parse.unquote(data_url)
+        if player_data.get('encrypt') == 2:
```
Comment (bashonly) [inline yt_dlp/extractor/duboku.py:136]:
```suggestion
        player_encrypt = player_data.get('encrypt')
        if player_encrypt == 1:
            data_url = urllib.parse.unquote(data_url)
        elif player_encrypt == 2:
```

**Review by bashonly — APPROVED** [review]

**Comment by dirkf** [comment]
Yes, the same tests that failed [before](https://github.com/yt-dlp/yt-dlp/pull/9161#issuecomment-1935193970) are passing now.

**Pull request merged**